### PR TITLE
Remove password authentication code.

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: components/idp.js [hunk 1]
```diff
@@ -45,7 +45,11 @@ module.config(function($routeProvider, routeResolverProvider) {
     // if route requires a session, redirect to login
     if($route.current.session === 'required') {
       // FIXME: use $location only once any SPA state issues are resolved
```
Comment (dlongley) [inline components/idp.js]: I see we're trying out `$location` ... are the SPA state issues resolved?

File: components/idp.js [hunk 2]
```diff
@@ -35,7 +35,7 @@ module.config(function($routeProvider, routeResolverProvider) {
   routeResolverProvider.add('bedrock-idp', 'session', resolve);
 
   /* @ngInject */
-  function resolve($window, $route) {
+  function resolve($location, $window, $route) {
```
Comment (dlongley) [inline components/idp.js]: Can we remove `$window` injectable?

File: components/idp.js [hunk 3]
```diff
@@ -111,7 +115,9 @@ module.config(function($routeProvider, routeResolverProvider) {
 });
 
 /* @ngInject */
-module.run(function($location, $rootScope, $route, $window, config, util) {
+module.run(function(
+  $location, $rootScope, $route, $window, brAuthnService, config, util) {
```
Comment (dlongley) [inline components/idp.js]: Can we remove `$window` injectable?

File: lib/auth.js [hunk 4]
```diff
@@ -245,57 +168,6 @@ function addViewVars(req, vars, callback) {
...
- *
- * @param req the request.
- * @param res the response.
- * @param next the next route handler.
- * @param callback(err, user, choice) called once the operation completes with
- *          the `user` that was logged in or false if there were multiple
- *          choices of users to log in and `choice` will contain the
- *          email address used and a map of identityId => identities that match.
- */
-api.login = function(req, res, next, callback) {
```
Comment (dlongley) [inline lib/auth.js]: Is this used anywhere? Can we avoid a major version bump when removing it?

File: lib/auth.js [hunk 5]
```diff
@@ -407,7 +244,9 @@ api.createIdentity = function(actor, identity, callback) {
           return callback(err);
         }
         // return unhashed passcode in record
-        record.identity.sysPasscode = results.generatePasscode;
+        // FIXME: is this necessary.  If so, a new solution will need to be
+        // implemented since passcode generation is happening elsewhere.
+        // record.identity.sysPasscode = results.generatePasscode;
```
Comment (dlongley) [inline lib/auth.js]: I think this was needed for the frontend or for an email that gets sent when the identity is created?

File: lib/auth.js [hunk 6]
```diff
@@ -417,22 +256,24 @@ api.createIdentity = function(actor, identity, callback) {
...
+ * Gets an IdentityID from the URL parameters in the given request. In the
+ * common case, the authenticated user for the request can be used to obtain
+ * the ID without an extra database hit.
  *
- * @param email the email address.
- * @param callback(err, identityIds) called once the operation completes.
+ * @param req the request.
+ * @param callback(err, identityId) called once the operation completes.
  */
-api.resolveEmail = function(email, callback) {
```
Comment (dlongley) [inline lib/auth.js]: We're getting rid of a lot of API functions -- can we deprecate them instead? Can these become wrappers around something from `bedrock-authn-password`? Are we going to require a database migration for these changes? If so, we need to try really hard to engineer around not having to do that.

File: lib/auth.js [hunk 7]
```diff
@@ -417,22 +256,24 @@ api.createIdentity = function(actor, identity, callback) {
...
-    if(records) {
-      records.forEach(function(record, i) {
-        records[i] = record.identity.id;
-      });
-    }
-    callback(err, records);
-  });
+api.getIdentityIdFromUrl = function(req, callback) {
+  // TODO: determine if this function's usage anywhere leaks DID information
+  // through its slug
```
Comment (dlongley) [inline lib/auth.js]: This sounds important, can you elaborate?

File: lib/auth.js [hunk 8]
```diff
@@ -417,22 +256,24 @@ api.createIdentity = function(actor, identity, callback) {
...
+
+  // if slug from URL matches authenticated user (common case), then
+  // use its ID
+  var slug = req.params.identity;
+  if(req.isAuthenticated() && req.user && req.user.identity.sysSlug === slug) {
```
Comment (dlongley) [inline lib/auth.js]: `req.isAuthenticated()` returns true when `req.user` is truthy -- so we don't need the additional `&& req.user` check.

File: lib/auth.js (hunk 6 above)
Comment (dlongley) [inline lib/auth.js]: To clarify, if these functions are never exposed outside of `bedrock-idp` we don't need to keep them available as deprecated. However, if they are -- we'll want to see if we can easily preserve them.

File: lib/auth.js (hunk 7 above)
Comment (mattcollier) [inline lib/auth.js]: This is your comment from: https://github.com/digitalbazaar/bedrock-idp/commit/f7c61ed48070149e0aa80bf0b478faa8f21561b3

Comment (dlongley) [inline lib/auth.js]: Ok, yeah, I thought it might be. I should have been more clear the first time!

Comment (dlongley) [inline lib/auth.js]: Nothing to do here. Just leave it.

File: bedrock/bower.json [hunk 9]
```diff
@@ -16,36 +16,17 @@
   "dependencies": {
     "bedrock-angular": "^2.0.0",
     "bedrock-angular-alert": "^2.0.0",
-    "bedrock-angular-credential": "digitalbazaar/bedrock-angular-credential#master",
-    "bedrock-angular-identity-composer": "digitalbazaar/bedrock-angular-identity-composer#master",
+    "bedrock-angular-credential": "^1.0.0",
+    "bedrock-angular-identity-composer": "^1.0.0",
     "bedrock-angular-form": "^2.0.0",
-    "bedrock-angular-key": "digitalbazaar/bedrock-angular-key#master",
+    "bedrock-angular-key": "^1.0.0",
```
Comment (dlongley) [inline bedrock/bower.json]: This is outdated now. All of these versions probably need to be checked/updated to latest.

File: bower.json [hunk 10]
```diff
@@ -24,14 +24,16 @@
...
     "bedrock-angular-form": "^2.0.0",
     "bedrock-angular-identity-composer": "^1.0.0",
     "bedrock-angular-key": "^1.0.0",
     "bedrock-angular-messages": "^1.0.0",
     "bedrock-angular-modal": "^3.0.0",
-    "bedrock-angular-navbar": "^2.0.0",
+    "bedrock-angular-navbar": "^3.0.0",
     "bedrock-angular-resolver": "^2.0.0",
     "bedrock-angular-resource": "^2.0.0",
     "bedrock-angular-session": "^2.0.0",
```
Comment (dlongley) [inline bower.json]: Ensure all versions are up-to-date.

File: components/login/login-controller.js [hunk 11]
```diff
@@ -25,7 +25,7 @@ function factory($scope, $http, $compile, $window, brAlertService, config) {
     // do login
     brAlertService.clearFeedback();
     self.loading = true;
-    Promise.resolve($http.post('/session/login', {
+    Promise.resolve($http.post('/authn-password/login', {
```
Comment (dlongley) [inline components/login/login-controller.js]: Add TODO to make configurable.

File: components/login/login.js [hunk 12]
```diff
@@ -26,10 +20,7 @@ define([
 var module = angular.module('bedrock-idp.login', []);
 
 module.controller(loginController);
-module.directive(loginModalDirective);
-module.directive(idpLoginDirective);
 module.directive(idpJoinDirective);
```
Comment (dlongley) [inline components/login/login.js]: Add an issue and assign to me for switching things over to the component API.

File: schemas/passcode.js [hunk 13]
```diff
@@ -8,10 +8,10 @@ var schema = {
...
-  minLength: 8,
-  maxLength: 8,
+  minLength: 40,
+  maxLength: 40,
   errors: {
-    invalid: 'The passcode must be 8 characters in length.',
+    invalid: 'The passcode must be 40 characters in length.',
     missing: 'Please enter a passcode.',
     masked: true
   }
```
Comment (dlongley) [inline schemas/passcode.js]: Is passcode schema used at all in idp directly now? Or is it just used via bedrock-authn-password? (If the latter, remove here).

File: schemas/password.js [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2012-2015 Digital Bazaar, Inc. All rights reserved.
+ * Copyright (c) 2012-2016 Digital Bazaar, Inc. All rights reserved.
  */
 var bedrock = require('bedrock');
```
Comment (dlongley) [inline schemas/password.js]: Can we remove this schema from here now as well?

File: schemas/services.session.js [hunk 15]
```diff
@@ -1,126 +0,0 @@
...
-module.exports.getPasscodeQuery = function() {
-  return getPasscodeQuery;
-};
-module.exports.postPasscode = function() {
-  return postPasscode;
-};
-module.exports.postPassword = function() {
-  return postPassword;
-};
-module.exports.postPasswordReset = function() {
```
Comment (dlongley) [inline schemas/services.session.js]: Does bedrock-authn-password provide password-reset/send passcode ability? (We don't completely lose this capability by removing this code, right?)

**Comment by dlongley** [comment]
Some comments/questions here to address. Nearly ready to merge, just need to hash those out first and make sure we're not losing some core required functionality.

File: lib/auth.js (hunk 6 above)
Comment (dlongley) [inline lib/auth.js]: This is now in `bedrock-authn-password` -- let's mark it deprecated and call into that module and make this a simple alias.

File: lib/auth.js [hunk 16]
```diff
@@ -478,52 +304,6 @@ api.resolveIdentitySlug = function(name, options, callback) {
 };
 
 /**
- * Gets the Identity IDs that match the given identifier. The identifier
- * can be an Identity ID, an Identity slug, or an email address.
- *
- * @param identifier the identifier to resolve.
- * @param callback(err, identityIds) called once the operation completes.
- */
-api.resolveIdentityIdentifier = function(identifier, callback) {
```
Comment (dlongley) [inline lib/auth.js]: This is now in `bedrock-authn-password` -- let's mark it deprecated and call into that module and make this a simple alias.

File: lib/auth.js [hunk 17]
```diff
@@ -561,241 +341,6 @@ api.setEmail = function(actor, id, email, callback) {
...
-
-/**
- * Verifies the Identity's passcode against the stored passcode and sets
- * the Identity's email address as verified upon success.
- *
- * @param actor the Identity performing the action.
- * @param identity the Identity with the passcode to verify.
- * @param callback(err, verified) called once the operation completes.
- */
-api.verifyEmail = function(actor, identity, callback) {
```
Comment (dlongley) [inline lib/auth.js]: This is now in `bedrock-authn-password` -- let's mark it deprecated and call into that module and make this a simple alias.

File: lib/auth.js [hunk 18]
```diff
@@ -561,241 +341,6 @@ api.setEmail = function(actor, id, email, callback) {
...
- * Sends an Identity's or multiple Identity's passcodes to their contact point
- * (eg: email address). The Identities must all have the same contact point and
- * must be populated.
- *
- * @param identities the Identities to send the passcode to.
- * @param usage 'reset' if the passcode is for resetting a password,
- *          'verify' if it is for verifying an email address/contact point.
- * @param callback(err) called once the operation completes.
- */
-api.sendPasscodes = function(identities, usage, callback) {
```
Comment (dlongley) [inline lib/auth.js]: This is now in `bedrock-authn-password` -- let's mark it deprecated and call into that module and make this a simple alias.

File: components/login/login-controller.js (hunk 11 above)
Comment (dlongley) [inline components/login/login-controller.js]: Should this be `/authn/password/login`?

**Comment by dlongley** [comment]
A couple of comments to address, then LGTM, merge.

**Comment by dlongley** [comment]
We'll need a semver major release ... which I think will be 5.0.0.

**Pull request merged**